dali/pipeline/data/tensor_vector.cc

```diff
@@ -12,268 +12,494 @@
+  size_t size = type_info().size();
+
+  for (int i = 0; i < shape_.size(); ++i) {
+    if (base_ptr != tensors_[i].raw_data()) {
+      return false;
     }
+    base_ptr += shape_[i].num_elements() * size;
   }
+  return true;
+}
 
-  other.views_count_ = 0;
-  other.curr_num_tensors_ = 0;
-  other.tensors_.clear();
-  other.sample_dim_ = -1;
+
+template <typename Backend>
+bool TensorVector<Backend>::IsDenseTensor() const {
+  return IsContiguous() && is_uniform(shape());
 }
 
+
+template <typename Backend>
+Tensor<Backend> TensorVector<Backend>::AsReshapedTensor(const TensorShape<> &new_shape) {
+  DALI_ENFORCE(num_samples() > 0,
+               "To create a view Tensor, the batch must have at least 1 element.");
+  DALI_ENFORCE(IsValidType(type()),
+               "To create a view Tensor, the batch must have a valid data type.");
+  DALI_ENFORCE(
+      shape().num_elements() == new_shape.num_elements(),
+      make_string("To create a view Tensor, requested shape need to have the same volume as the "
+                  "batch, requested: ",
+                  new_shape.num_elements(), " expected: ", shape().num_elements()));
+  Tensor<Backend> result;
+  result.ShareData(contiguous_buffer_.get_data_ptr(), contiguous_buffer_.capacity(),
+                   contiguous_buffer_.is_pinned(), new_shape, type(), device_id(), order());
+  auto result_layout = GetLayout();
+  if (!GetLayout().empty()) {
+    result_layout = TensorLayout("N") + result_layout;
+  }
+  result.SetLayout(result_layout);
+  return result;
+}
+
+
+template <typename Backend>
+Tensor<Backend> TensorVector<Backend>::AsTensor() {
+  DALI_ENFORCE(IsDenseTensor(),
+               "The batch must be representable tensor - it must has uniform shape and be "
+               "allocated in contiguous memory.");
+  DALI_ENFORCE(shape().num_samples() > 0,
+               "To create a view Tensor, the batch must have at least 1 element.");
+  return AsReshapedTensor(shape_cat(shape().num_samples(), shape()[0]));
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::VerifySampleShareConformance(DALIDataType type, int sample_dim,
+                                                         TensorLayout layout, bool pinned,
+                                                         AccessOrder order, int device_id,
+                                                         const std::string &error_suffix) {
+  // Checks in the order of class members
+  DALI_ENFORCE(this->type() == type,
+               make_string("Sample must have the same type as the target batch, current: ",
+                           this->type(), ", new: ", type, error_suffix));
+
+  DALI_ENFORCE(this->sample_dim() == sample_dim,
+               make_string("Sample must have the same sample dim as the target batch, current: ",
+                           this->sample_dim(), ", new: ", sample_dim, error_suffix));
+
+  DALI_ENFORCE(this->GetLayout() == layout,
```

Comment:
I guess we could accept a sample without a layout and adopt whatever layout is currently set.

Reply:
Done.